// dht/lib/counter.js
//服务端存储
//exports
var util = require('./util');
var http = require('http');
var url=require('url');
var qs=require('querystring');

function Counter(delay){
    console.log('counter init');
    if(!delay){
        delay = 30 * 1000;
    }

    var self = this;
    this.dataQueue = {
        'counter': {
            'send':0,
            'get_peers': 0,
            'find_node': 0,
            'ping': 0,
            'announce_peer': 0
        },
        'hash': []
    };

    this.recordHash = function(hash){
        self.dataQueue.hash.push(hash);
    }
    this.recordCount = function(type){
        if(self.dataQueue.counter[type]){
            self.dataQueue.counter[type]++;
        }else{
            self.dataQueue.counter[type] = 1;
        }
    }
    setInterval(function(){
        _sendHash(self.dataQueue.hash);

        self.dataQueue.hash = [];
        for (var type in self.dataQueue.counter) {
            self.dataQueue.counter[type] = 0;
        }
    }, delay);

    //废弃单独计数
    function _sendCounter(type, num, callback){
        if(!num)return;
        util.debug('send: counter' + type + num);
        http.get(url.parse("http://shenmaohelp.sinaapp.com/?m=Dht&a=counter&type=" + type + "&num=" + num), function(res) {
            util.debug("counter: " + res.statusCode);
            callback && callback();
        }).on('error', function(e) {
                util.debug("counter: " + e.message);
        });
    }
    function _sendHash(hashes, callback){
        if(!hashes || !hashes.length) return;

        util.debug('send: hash' + self.dataQueue.hash.length);
        var hash = hashes.join(',');

        var data = {
            info_hash: hash,
            count: JSON.stringify(self.dataQueue.counter)
        };
        data = qs.stringify(data);
        var options = {
            hostname: 'shenmaohelp.sinaapp.com',
            port: 80,
            path: '/?m=Dht&a=addHash',
            method: 'POST',
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded',
                'Content-Length': data.length
            }
        }
        var req = http.request(options, function(res) {
            var _data='';
            res.on('data', function(chunk){
                _data += chunk;
            });
            res.on('end', function(){
//                console.log("\n--->>\nresult:",_data)
            });
        });
        req.write(data);
        req.end();
    }
}
exports.Counter = Counter;
