// dht/lib/storetorrent.js

var mongodb = require("mongodb");

function StoreTorrent(){

    var server = new mongodb.Server('localhost',27017,{auto_reconnect:true}, 10);

    var db = new mongodb.Db("warm",server);


    var self = this;
    db.open(function(err,db){
        if(err){
            console.log(err);
        }
        if(!err){
            console.log("we are connected!");
            db.collection('torrents', function(err,collection){
                self.collection = collection;
//                insertTorrent(12445);
//                getTopTorrent(5);
            });
        }
    });

    this.insertTorrent = function(info_hash){
        if(!self.collection){
            return;
        }
        self.collection.update({'info_hash': info_hash}, {$inc:{'num': 1}} ,{upsert:true, w: 1},function(err,result){
            if(err) throw err;
//            console.log('insert info_hash ' + info_hash);
        });

    }
    this.getTopTorrent = function(num, callback){
        if(!self.collection){
            return;
        }
        self.collection.find({
                'is_read': {$ne: 1}
            },
            {
                sort:{
                    'num': -1
                },
                limit: num,
                fields: {'info_hash': 1, _id : 0}
            }).toArray(function(err, docs) {
                var infos = [];
                for(var i = 0; i < docs.length; i++){
                    infos.push(docs[i]['info_hash']);
                    self.collection.update({'info_hash': docs[i]['info_hash']}, {$set:{'is_read': 1}});
                }
                console.log(infos);
                callback && callback(infos);
            });
    }
}

exports.storetorrent = new StoreTorrent();
